File: problems/CSES/Dynamic_Programming/removing_digits/removing_digits.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;

    const int INF = 1e9;

    vector<int> dp(n + 1, INF);
    dp[0] = 0;

    for (int number = 1; number <= n + 1; number++) {
        int value = number;

        while (value > 0) {
            int digit = value % 10;
            value /= 10;

            if (digit != 0) {
                dp[number] = min(dp[number], dp[number - digit] + 1);
            }
        }
    }

    cout << dp[n] << "\n";

    return 0;
}

File: problems/CSES/Dynamic_Programming/removing_digits/removing_digits.py
import sys
# sys.setrecursionlimit(200000)

input = sys.stdin.readline

def main():
    n = int(input())

    dp = [0] + [10**9] * n

    for number in range(1, n + 1):
        value = number

        while value > 0:
            digit = value % 10
            value //= 10

            if digit != 0:
                dp[number] = min(
                    dp[number],
                    dp[number - digit] + 1
                )

    print(dp[n])

main()
